src/com/ableton/PythonBuilder.groovy

```diff
@@ -0,0 +1,101 @@
+package com.ableton
+
+
+class PythonBuilder implements Serializable {
+  @SuppressWarnings('FieldTypeRequired')
+  def script = null
+  /**
+   * Python major version to download (required value). Example: '3.7.0'
+   */
+  String version = null
+
+  /**
+   * Mirror to use when downloading Python.
+   */
+  String downloadMirror = 'https://www.python.org/ftp/python'
+  /**
+   * Method to use to download Python. Currently supported methods include:
+   * - 'curl': Requires curl to be installed on the build node.
+   * - 'httpRequest': Requires the Jenkins HTTP Request Plugin to be installed (default).
+   */
+  String downloadWith = 'httpRequest'
+  /**
+   * Number of parallel jobs to use when building Python (ie, `make -jN`). Default: '1'
+   */
+  String makeJobs = '1'
+  /**
+   * Python version revision, optional. This value is appended to the major version, but
+   * because of the way that Python's downloads are organized, this value must be separate
+   * from the major version. So for Python 3.7.0a4, use '3.7.0' for the version and 'a4'
+   * for the revision.
+   */
+  String revision = ''
+
+  String install() {
+    // Sanity check required values
+    assert script
+    assert version
+
+    if (!script.isUnix()) {
+      throw new UnsupportedOperationException('This function is only supported on Unix')
+    }
```

Comment:
What is only supported on Unix?
- Calling `download()`? Then `download()` itself should complain.
- Calling `extract()`? Then `extract()` itself should complain.
- Calling `build()`? Then `build()` itself should complain.

Reply:
All of those functions will only work on Unix. 😄 This is mostly due to assumptions about path separators, but there are other things which could easily go wrong on Windows which I have not tested. For example, we call `sh` instead of `bat` which won't work on Windows hosts which lack a Unix toolchain. Our Windows build hosts have that, but we can't make any such assumptions for others (and this is a public library).

I would prefer to exclude that platform until we have a need to support it, rather than invest the time in making this code cross-platform without any real need to do so.

Am I being too lazy here? 😅

Reply:
You are not. I am fine with excluding Windows. But let's exclude it properly.

Reply:
How should we do that, then?

Reply:
Can we add a ctor and throw in the ctor?

Reply:
No, I would prefer not to do that. I have really come around to the idea of not defining ctors and letting Groovy autogenerate them, ever since we decided to go that route with `DevToolsProject`.

Let me explain the way the language works, because it might not be obvious here (it wasn't to me until I read up on it). By not defining a ctor but instead setting default values for fields, you can use "named arguments" when creating an object like so:

```groovy
class Foo {
  String bar = 'bar'
  String baz = 'baz'
}

Foo foo = new Foo(bar: 'not-bar', baz: 'not-baz')
```

I think that we can both agree that we like this syntax. 😁 So the way that works behind the scenes is that the Groovy generates a bunch of default ctors, including one which takes a `Map` and (presumably) uses all sorts of reflection to assign the values to their respective fields. The call to `new Foo` converts the args to a map behind the scenes (note that we *don't* need to use `[:]` when calling `new Foo([bar: 'not-bar' ...])`). So the point is, once you have a custom ctor, you can't create objects using the named argument syntax. In other words, this will break:

```groovy
class Foo {
  String bar
  String baz

  Foo(String bar, String baz) {
    this.bar = bar
    this.baz = baz
  }
}

// Error: Will throw groovy.lang.GroovyRuntimeException
// With a message like "Could not find matching constructor for: Foo(java.util.LinkedHashMap)"
Foo foo = new Foo(bar: 'not-bar', baz: 'not-baz')
```

So basically, named parameters aren't really named parameters. As such, I'd prefer to use the calling syntax where we can pass named parameters rather than bloat the ctor and `install` with a bunch of arguments with default values.

Reply:
Hmm, now that I think about it, I guess we *could* also provide a custom ctor which also takes a `Map`, but I find this code to be rather ugly:

```groovy
class Foo {
  String bar
  String baz

  Foo(Map args) {
    bar = args?.bar ?: 'bar'
    baz = args?.baz ?: 'baz'
  }
}
```

Regardless, not generating the ctor and letting Groovy do its thing is definitely the "Groovy way".

Reply:
Then let's not add a ctor :+1: 
I've just realized that `install` is the only public method :sweat_smile: Forget about what I said, sorry for the noise :retreat:

Reply:
@ala-ableton Sounds good, what changes do I need to make to this PR then? I think I've addressed your other concerns, no?